File: cane-cpp/cane-core/src/constraints/transform.cpp
#include "solvers.hpp"

namespace cane::constraints {
namespace {
using Properties = std::array<float, 6>;
float& component(BoneLocal& value, std::size_t property) {
    switch (property) { case 0: return value.rotation_degrees; case 1: return value.x; case 2: return value.y; case 3: return value.scale_x; case 4: return value.scale_y; default: return value.shear_y_degrees; }
}
float positive(float value) { const auto r = std::fmod(value, 360.0f); return r < 0 ? r + 360.0f : r; }
Properties source_values(const Properties& offsets, BoneLocal local, const Affine& world, bool local_source) {
    Properties source{};
    if (local_source) { for (std::size_t i = 0; i < source.size(); ++i) source[i] = component(local, i) + offsets[i]; return source; }
    const auto p = world.transform({offsets[1], offsets[2]}); const auto reflection = world.determinant() < 0 ? -1.0f : 1.0f;
    return {positive(x_angle(world) + offsets[0] * reflection), p.x, p.y, x_scale(world) + offsets[3], y_scale(world) + offsets[4], constraint_degrees(y_angle(world) - x_angle(world) - 90 + offsets[5])};
}
BoneLocal legacy_local(const detail::TransformDefinition& d, const Properties& offsets, BoneLocal current, BoneLocal target, const Properties& mix) {
    auto next = current;
    for (std::size_t i = 0; i < mix.size(); ++i) {
        const auto before = component(current, i), source = component(target, i), offset = offsets[i], amount = mix[i];
        if (amount == 0) continue;
        if (i == 3 || i == 4) component(next, i) = d.relative ? before * ((source - 1 + offset) * amount + 1) : before == 0 ? before : before + (source - before + offset) * amount;
        else if (i == 0 || i == 5) component(next, i) = constraint_degrees(before + (d.relative ? source + offset : constraint_degrees(source + offset - before)) * amount);
        else component(next, i) = before + (source + offset - (d.relative ? 0 : before)) * amount;
    }
    return next;
}
Affine shear_world(Affine current, float delta, float mix) {
    const auto radians = (y_angle(current) + constraint_degrees(delta) * mix) * degrees_to_radians;
    const auto length = y_scale(current);
    current.c = static_cast<float>(std::cos(static_cast<double>(radians))) * length;
    current.d = static_cast<float>(std::sin(static_cast<double>(radians))) * length; return current;
}
Affine legacy_world(const detail::TransformDefinition& d, const Properties& offsets, Affine current, const Affine& target, const Properties& mix) {
    const auto reflection = target.determinant() < 0 ? -1.0f : 1.0f;
    if (mix[0] != 0) current = rotate_axes(current, constraint_degrees(x_angle(target) - (d.relative ? 0 : x_angle(current)) + offsets[0] * reflection) * mix[0]);
    if (mix[1] != 0 || mix[2] != 0) {
        const auto point = target.transform({offsets[1], offsets[2]});
        if (mix[1] != 0) current.tx += (point.x - (d.relative ? 0 : current.tx)) * mix[1];
        if (mix[2] != 0) current.ty += (point.y - (d.relative ? 0 : current.ty)) * mix[2];
    }
    const auto sx = x_scale(current), sy = y_scale(current); float fx = 1, fy = 1;
    if (mix[3] != 0 && sx > matrix_epsilon) fx = d.relative ? (x_scale(target) - 1 + offsets[3]) * mix[3] + 1 : (sx + (x_scale(target) - sx + offsets[3]) * mix[3]) / sx;
    if (mix[4] != 0 && sy > matrix_epsilon) fy = d.relative ? (y_scale(target) - 1 + offsets[4]) * mix[4] + 1 : (sy + (y_scale(target) - sy + offsets[4]) * mix[4]) / sy;
    current.a *= fx; current.b *= fx; current.c *= fy; current.d *= fy;
    if (mix[5] != 0) current = shear_world(current, constraint_degrees(y_angle(target) - x_angle(target)) - (d.relative ? 90 : constraint_degrees(y_angle(current) - x_angle(current))) + offsets[5] * reflection, mix[5]);
    return current;
}
Affine mapped_world(Affine current, std::size_t property, float value, float amount, bool relative) {
    if (property == 0) return rotate_axes(current, constraint_degrees(relative ? value : value - x_angle(current)) * amount);
    if (property == 1) { current.tx += (relative ? value : value - current.tx) * amount; return current; }
    if (property == 2) { current.ty += (relative ? value : value - current.ty) * amount; return current; }
    if (property == 3 || property == 4) {
        const auto length = property == 3 ? x_scale(current) : y_scale(current); if (length <= matrix_epsilon) return current;
        const auto factor = relative ? 1 + (value - 1) * amount : 1 + (value - length) * amount / length;
        if (property == 3) { current.a *= factor; current.b *= factor; } else { current.c *= factor; current.d *= factor; }
        return current;
    }
    return shear_world(current, relative ? value : value + 90 - constraint_degrees(y_angle(current) - x_angle(current)), amount);
}
void project(const detail::TransformDefinition& d, const Properties& offsets, const Properties& source, const BoneLocal& target_local, const Affine& target_world,
             const Properties& mixes, bool local_target, BoneLocal& local, Affine& world) {
    if (!d.mapping) {
        if (local_target) local = legacy_local(d, offsets, local, target_local, mixes); else world = legacy_world(d, offsets, world, target_world, mixes);
        return;
    }
    for (const auto& from : d.mapping->properties) {
        const auto s = source[static_cast<std::size_t>(from.property)] - from.offset;
        for (const auto& to : from.targets) {
            const auto property = static_cast<std::size_t>(to.property); const auto amount = mixes[property]; if (amount == 0) continue;
            auto value = to.offset + s * to.scale;
            if (d.mapping->clamp) value = std::clamp(value, std::min(to.offset, to.max), std::max(to.offset, to.max));
            if (local_target) {
                auto& current = component(local, property);
                if (property == 3 || property == 4) { if (d.relative) current *= 1 + (value - 1) * amount; else if (current != 0) current += (value - current) * amount; }
                else current += (d.relative ? value : value - current) * amount;
            } else world = mapped_world(world, property, value, amount, d.relative);
        }
    }
}
void maximum(std::optional<float>& output, float value) {
    if (!std::isfinite(value)) throw Error(ErrorCode::non_finite, "apply", "Transform diagnostic is non-finite.", "constraints");
    output = output ? std::max(*output, value) : value;
}
}
void solve_transform(Context& context, std::size_t index) {
    auto& pose = context.pose; pose.diagnostics.at(index) = std::monostate{};
    const auto& d = std::get<detail::TransformDefinition>(pose.model().constraints.at(index).definition);
    if (!pose.active_constraints[index] || !pose.active_bones[d.target]) return;
    Properties mixes{}, full{}; bool active = false;
    for (std::size_t i = 0; i < mixes.size(); ++i) {
        mixes[i] = pose.constraints[index].scalars[static_cast<std::size_t>(animation::ConstraintProperty::mix_rotate) + i];
        full[i] = mixes[i] == 0 ? 0.0f : 1.0f; active = active || mixes[i] != 0;
    }
    if (!active) return;
    const auto world = pose.world[d.target]; const auto local = pose.locals[d.target];
    const auto source = d.mapping ? source_values(pose.constraints[index].transform_offsets, local, world, d.mapping->local_source) : Properties{};
    const bool local_target = d.mapping ? d.mapping->local_target : d.local;
    TransformDiagnostic diagnostic; diagnostic.mixes = mixes;
    for (const auto bone : d.bones) {
        if (bone == d.target) continue;
        auto next_local = pose.locals[bone], ideal_local = next_local; auto next_world = pose.world[bone], ideal_world = next_world;
        project(d, pose.constraints[index].transform_offsets, source, local, world, mixes, local_target, next_local, next_world);
        project(d, pose.constraints[index].transform_offsets, source, local, world, full, local_target, ideal_local, ideal_world);
        if (local_target) { ideal_world = context.compose(bone, ideal_local); context.write_local(bone, next_local); }
        else context.write_world(bone, next_world);
        const auto& actual = pose.world[bone]; ++diagnostic.driven;
        if (mixes[1] != 0 || mixes[2] != 0) maximum(diagnostic.translation_residual, cane::hypot(mixes[1] == 0 ? 0 : actual.tx - ideal_world.tx, mixes[2] == 0 ? 0 : actual.ty - ideal_world.ty));
        if (mixes[0] != 0) maximum(diagnostic.rotation_residual, std::abs(constraint_degrees(x_angle(actual) - x_angle(ideal_world))));
        if (mixes[3] != 0 || mixes[4] != 0) maximum(diagnostic.scale_residual, std::max(mixes[3] == 0 ? 0 : std::abs(x_scale(actual) - x_scale(ideal_world)), mixes[4] == 0 ? 0 : std::abs(y_scale(actual) - y_scale(ideal_world))));
        if (mixes[5] != 0) maximum(diagnostic.shear_residual, std::abs(constraint_degrees(world_shear_y(actual) - world_shear_y(ideal_world))));
    }
    pose.diagnostics[index] = diagnostic;
}
std::optional<std::array<float, 6>> matched_transform_offsets(const Context& context, std::size_t index, ErrorCode* failure_code) {
    if (failure_code) *failure_code = ErrorCode::invalid_state;
    const auto& pose = context.pose; const auto& d = std::get<detail::TransformDefinition>(pose.model().constraints.at(index).definition);
    bool local_source = d.local, local_target = d.local;
    if (d.mapping) {
        const auto& mapping = *d.mapping; if (mapping.clamp || mapping.properties.size() != 6) return {};
        std::array<bool, 6> seen{};
        for (const auto& source : mapping.properties) {
            const auto p = static_cast<std::size_t>(source.property);
            if (seen[p] || source.offset != 0 || source.targets.size() != 1) return {};
            const auto& target = source.targets[0]; if (target.property != source.property || target.offset != 0 || target.scale != 1) return {};
            seen[p] = true;
        }
        local_source = mapping.local_source; local_target = mapping.local_target;
    }
    const auto values = [&](std::size_t bone, bool local) -> Properties {
        if (local) { auto p = pose.locals[bone]; return {p.rotation_degrees, p.x, p.y, p.scale_x, p.scale_y, p.shear_y_degrees}; }
        const auto& world = pose.world[bone]; return {positive(x_angle(world)), world.tx, world.ty, x_scale(world), y_scale(world), world_shear_y(world)};
    };
    const auto desired = d.relative ? Properties{0, 0, 0, 1, 1, 0} : values(d.bones.front(), local_target);
    const auto source = values(d.target, local_source); Properties offsets{};
    offsets[0] = constraint_degrees(desired[0] - source[0]);
    if (local_source) { offsets[1] = desired[1] - source[1]; offsets[2] = desired[2] - source[2]; }
    else {
        const auto& world = pose.world[d.target]; const auto inverse = world.inverse(matrix_epsilon); if (!inverse) return {};
        const auto position = inverse->transform({desired[1], desired[2]}); offsets[1] = position.x; offsets[2] = position.y;
        if (world.determinant() < 0) offsets[0] = -offsets[0];
    }
    offsets[3] = desired[3] - source[3]; offsets[4] = desired[4] - source[4]; offsets[5] = constraint_degrees(desired[5] - source[5]);
    for (const auto value : offsets) if (!std::isfinite(value)) { if (failure_code) *failure_code = ErrorCode::non_finite; return {}; }
    return offsets;
}
}

File: cane-csharp/cane-core/Runtime/Constraints/Transform.cs
using System;
using Cane.Animation;
using Cane.Format;

namespace Cane.Constraints
{
    internal static class Transform
    {
        private static readonly string[] Properties = { "rotate", "x", "y", "scaleX", "scaleY", "shearY" };
        private static readonly string[] Mixes = { "mixRotate", "mixX", "mixY", "mixScaleX", "mixScaleY", "mixShearY" };
        private static readonly string[] Offsets = { "rotation", "x", "y", "scaleX", "scaleY", "shearY" };
        private static readonly int[] Components = { 2, 0, 1, 3, 4, 6 };
        internal static void Solve(Pose pose, Json c, Affine root)
        {
            var mixes = new float[6]; bool active = false;
            for (int i = 0; i < 6; i++) { mixes[i] = c.F(Mixes[i], 1); active |= mixes[i] != 0; } if (!active) return;
            int target = pose.Data.BoneIndex[c.S("targetBoneId")]; Affine world = pose.World[target]; BoneLocal local = pose.Locals[target];
            Json mapping = c["mapping"]; bool relative = c.B("relative"); float[]? source = mapping.IsNull ? null : Source(c, local, world, mapping.B("localSource"));
            var fullMixes = new float[6]; for (int i = 0; i < 6; i++) fullMixes[i] = mixes[i] == 0 ? 0 : 1;
            float? translationResidual = null, rotationResidual = null, scaleResidual = null, shearResidual = null; int driven = 0;
            foreach (Json id in c["boneIds"].Items)
            {
                int bone = pose.Data.BoneIndex[id.String]; if (bone == target) continue;
                BoneLocal nextLocal = pose.Locals[bone]; Affine nextWorld = pose.World[bone];
                BoneLocal idealLocal = nextLocal; Affine idealWorld = nextWorld;
                bool localTarget = mapping.IsNull ? c.B("local") : mapping.B("localTarget");
                Project(c, mapping, source, local, world, localTarget, mixes, relative, ref nextLocal, ref nextWorld);
                Project(c, mapping, source, local, world, localTarget, fullMixes, relative, ref idealLocal, ref idealWorld);
                if (localTarget)
                {
                    Affine parent = Solvers.Parent(pose, bone, root);
                    idealWorld = pose.Data.BoneData[bone].Parent < 0 ? parent * Affine.FromLocal(idealLocal) : Affine.Child(parent, idealLocal, pose.Modes[bone]);
                    Solvers.WriteLocal(pose, bone, nextLocal, root);
                }
                else Solvers.WriteWorld(pose, bone, nextWorld, root);
                Affine actual = pose.World[bone]; driven++;
                if (mixes[1] != 0 || mixes[2] != 0) Solvers.IncludeMax(ref translationResidual, Numeric.Hypot(mixes[1] == 0 ? 0 : actual.Tx - idealWorld.Tx, mixes[2] == 0 ? 0 : actual.Ty - idealWorld.Ty));
                if (mixes[0] != 0) Solvers.IncludeMax(ref rotationResidual, Math.Abs(Solvers.Degrees(Solvers.Angle(actual) - Solvers.Angle(idealWorld))));
                if (mixes[3] != 0 || mixes[4] != 0) Solvers.IncludeMax(ref scaleResidual, Math.Max(mixes[3] == 0 ? 0 : Math.Abs(Solvers.Sx(actual) - Solvers.Sx(idealWorld)), mixes[4] == 0 ? 0 : Math.Abs(Solvers.Sy(actual) - Solvers.Sy(idealWorld))));
                if (mixes[5] != 0) Solvers.IncludeMax(ref shearResidual, Math.Abs(Solvers.Degrees(Solvers.ShearY(actual) - Solvers.ShearY(idealWorld))));
            }
            pose.Diagnostics[c.S("id")] = new TransformConstraintDiagnostic(driven, mixes, translationResidual, rotationResidual, scaleResidual, shearResidual);
        }
        private static void Project(Json c, Json mapping, float[]? source, BoneLocal targetLocal, Affine targetWorld, bool localTarget, float[] mixes, bool relative, ref BoneLocal nextLocal, ref Affine nextWorld)
        {
            if (mapping.IsNull)
            {
                if (localTarget) nextLocal = LegacyLocal(c, nextLocal, targetLocal, mixes, relative);
                else nextWorld = LegacyWorld(c, nextWorld, targetWorld, mixes, relative);
                return;
            }
            foreach (Json from in mapping["properties"].Items)
            {
                float s = source![Array.IndexOf(Properties, from.S("property"))] - from.F("offset");
                foreach (Json to in from["targets"].Items)
                {
                    int prop = Array.IndexOf(Properties, to.S("property")); float amount = mixes[prop]; if (amount == 0) continue;
                    float offset = to.F("offset"), value = offset + s * to.F("scale", 1);
                    if (mapping.B("clamp")) value = Numeric.Clamp(value, Math.Min(offset, to.F("max", 1)), Math.Max(offset, to.F("max", 1)));
                    if (localTarget)
                    {
                        int component = Components[prop]; float current = nextLocal.Get(component), next = current;
                        if (prop == 3 || prop == 4) { if (relative) next = current * (1 + (value - 1) * amount); else if (current != 0) next = current + (value - current) * amount; }
                        else next = current + (relative ? value : value - current) * amount;
                        nextLocal = nextLocal.With(component, next);
                    }
                    else nextWorld = MappedWorld(nextWorld, prop, value, amount, relative);
                }
            }
        }
        internal static TransformConstraintOffsets Match(Pose pose, Json c)
        {
            const string op = "queryMatchedTransformConstraintOffsets"; string id = c.S("id"); Json mapping = c["mapping"];
            bool localSource = c.B("local"), localTarget = localSource;
            if (!mapping.IsNull)
            {
                if (!Canonical(mapping)) throw new RuntimeException(RuntimeErrorCode.InvalidState, op, "Match requires canonical property routing.", "mapping", id);
                localSource = mapping.B("localSource"); localTarget = mapping.B("localTarget");
            }
            int sourceBone = pose.Data.BoneIndex[c.S("targetBoneId")], constrained = pose.Data.BoneIndex[c["boneIds"][0].String];
            Affine sourceWorld = pose.World[sourceBone];
            float[] desired = c.B("relative") ? new float[] { 0, 0, 0, 1, 1, 0 } : MatchValues(pose, constrained, localTarget);
            float[] source = MatchValues(pose, sourceBone, localSource);
            float x, y, rotation = Solvers.Degrees(desired[0] - source[0]);
            if (localSource) { x = desired[1] - source[1]; y = desired[2] - source[2]; }
            else
            {
                float determinant = sourceWorld.Determinant;
                // Transform Match uses the constraint inverse contract, whose threshold differs from source-vertex inverse authoring.
                if (Math.Abs(determinant) <= Numeric.Epsilon || !Numeric.Finite(determinant)) throw new RuntimeException(RuntimeErrorCode.InvalidState, op, "Source transform has no constraint inverse.", "targetBoneId", id);
                float dx = desired[1] - sourceWorld.Tx, dy = desired[2] - sourceWorld.Ty;
                x = (sourceWorld.D * dx - sourceWorld.C * dy) / determinant; y = (-sourceWorld.B * dx + sourceWorld.A * dy) / determinant;
                if (determinant < 0) rotation = -rotation;
            }
            return new TransformConstraintOffsets(id, rotation, x, y, desired[3] - source[3], desired[4] - source[4], Solvers.Degrees(desired[5] - source[5]));
        }
        private static bool Canonical(Json mapping)
        {
            if (mapping.B("clamp") || mapping["properties"].Items.Count != 6) return false;
            var seen = new bool[6];
            foreach (Json from in mapping["properties"].Items)
            {
                int index = Array.IndexOf(Properties, from.S("property"));
                if (index < 0 || seen[index] || from.F("offset") != 0 || from["targets"].Items.Count != 1) return false;
                Json to = from["targets"][0]; if (to.S("property") != from.S("property") || to.F("offset") != 0 || to.F("scale", 1) != 1) return false;
                seen[index] = true;
            }
            return true;
        }
        private static float[] MatchValues(Pose pose, int bone, bool local)
        {
            if (local) { BoneLocal p = pose.Locals[bone]; return new[] { p.RotationDegrees, p.X, p.Y, p.ScaleX, p.ScaleY, p.ShearYDegrees }; }
            Affine world = pose.World[bone]; return new[] { Solvers.Positive(Solvers.Angle(world)), world.Tx, world.Ty, Solvers.Sx(world), Solvers.Sy(world), Solvers.ShearY(world) };
        }
        private static BoneLocal LegacyLocal(Json c, BoneLocal current, BoneLocal target, float[] mix, bool relative)
        {
            BoneLocal next = current;
            for (int i = 0; i < 6; i++)
            {
                float before = current.Get(Components[i]), source = target.Get(Components[i]), offset = c.F(Offsets[i]), amount = mix[i], result;
                if (i == 3 || i == 4)
                    result = relative ? before * ((source - 1 + offset) * amount + 1) : before == 0 || amount == 0 ? before : before + (source - before + offset) * amount;
                else if (i == 0 || i == 5)
                    result = Solvers.Degrees(before + (relative ? source + offset : Solvers.Degrees(source + offset - before)) * amount);
                else result = before + (source + offset - (relative ? 0 : before)) * amount;
                next = next.With(Components[i], result);
            }
            return next;
        }
        private static Affine LegacyWorld(Json c, Affine current, Affine target, float[] mix, bool relative)
        {
            float reflection = target.Determinant < 0 ? -1 : 1;
            if (mix[0] != 0) current = Solvers.Rotate(current, Solvers.Degrees(Solvers.Angle(target) - (relative ? 0 : Solvers.Angle(current)) + c.F("rotation") * reflection) * mix[0]);
            Point translated = target.Transform(c.F("x"), c.F("y"));
            current = new Affine(current.A, current.B, current.C, current.D,
                current.Tx + (translated.X - (relative ? 0 : current.Tx)) * mix[1], current.Ty + (translated.Y - (relative ? 0 : current.Ty)) * mix[2]);
            float sx = Solvers.Sx(current), sy = Solvers.Sy(current), fx = 1, fy = 1;
            if (mix[3] != 0 && sx > Numeric.Epsilon) fx = relative ? (Solvers.Sx(target) - 1 + c.F("scaleX")) * mix[3] + 1 : (sx + (Solvers.Sx(target) - sx + c.F("scaleX")) * mix[3]) / sx;
            if (mix[4] != 0 && sy > Numeric.Epsilon) fy = relative ? (Solvers.Sy(target) - 1 + c.F("scaleY")) * mix[4] + 1 : (sy + (Solvers.Sy(target) - sy + c.F("scaleY")) * mix[4]) / sy;
            current = new Affine(current.A * fx, current.B * fx, current.C * fy, current.D * fy, current.Tx, current.Ty);
            if (mix[5] != 0)
            {
                float y = Solvers.YAngle(current), targetDelta = Solvers.Degrees(Solvers.YAngle(target) - Solvers.Angle(target));
                float delta = targetDelta - (relative ? 90 : Solvers.Degrees(y - Solvers.Angle(current))) + c.F("shearY") * reflection;
                float angle = (y + Solvers.Degrees(delta) * mix[5]) * Numeric.DegToRad, length = Solvers.Sy(current);
                current = new Affine(current.A, current.B, Numeric.Cos(angle) * length, Numeric.Sin(angle) * length, current.Tx, current.Ty);
            }
            return current;
        }
        private static float[] Source(Json c, BoneLocal local, Affine world, bool localSource)
        {
            if (localSource) { var source = new float[6]; for (int i = 0; i < 6; i++) source[i] = local.Get(Components[i]) + c.F(Offsets[i]); return source; }
            Point p = world.Transform(c.F("x"), c.F("y")); float reflection = world.Determinant < 0 ? -1 : 1;
            return new[] { Solvers.Positive(Solvers.Angle(world) + c.F("rotation") * reflection), p.X, p.Y, Solvers.Sx(world) + c.F("scaleX"), Solvers.Sy(world) + c.F("scaleY"), Solvers.Degrees(Solvers.YAngle(world) - Solvers.Angle(world) - 90 + c.F("shearY")) };
        }
        private static Affine MappedWorld(Affine current, int property, float value, float amount, bool relative)
        {
            switch (property)
            {
                case 0: return Solvers.Rotate(current, Solvers.Degrees(relative ? value : value - Solvers.Angle(current)) * amount);
                case 1: return new Affine(current.A, current.B, current.C, current.D, current.Tx + (relative ? value : value - current.Tx) * amount, current.Ty);
                case 2: return new Affine(current.A, current.B, current.C, current.D, current.Tx, current.Ty + (relative ? value : value - current.Ty) * amount);
                case 3: case 4:
                    float length = property == 3 ? Solvers.Sx(current) : Solvers.Sy(current); if (length <= Numeric.Epsilon) return current;
                    float f = relative ? 1 + (value - 1) * amount : 1 + (value - length) * amount / length;
                    return property == 3 ? new Affine(current.A * f, current.B * f, current.C, current.D, current.Tx, current.Ty) : new Affine(current.A, current.B, current.C * f, current.D * f, current.Tx, current.Ty);
                default:
                    float y = Solvers.YAngle(current), delta = relative ? value : value + 90 - Solvers.Degrees(y - Solvers.Angle(current));
                    float angle = (y + Solvers.Degrees(delta) * amount) * Numeric.DegToRad, magnitude = Solvers.Sy(current);
                    return new Affine(current.A, current.B, Numeric.Cos(angle) * magnitude, Numeric.Sin(angle) * magnitude, current.Tx, current.Ty);
            }
        }
    }
}
